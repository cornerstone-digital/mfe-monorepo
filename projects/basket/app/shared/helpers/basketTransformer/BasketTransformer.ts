import getCookie from '@web-shop-core/helpers/getCookie'
import get from 'lodash/get'

import { sanitizeJson } from '@shared/helpers/transformUtils/transformUtils'
import sortByObjectProperty from '@helpers/sortByObjectProperty'
import { isUpgradeJourney } from '@helpers/typeCheck'
import { UserState } from '@pages/BasketPage/BasketPage.types'
import matchesConstCI from '@helpers/matchesConstCI'
import { BASKET_CONSTS } from '@constants'

export interface TransformedBasket extends BasketV2.Basket {
  isBusiness?: boolean
  isReviewMode?: boolean
  isUpgradeOrder?: boolean
  isBannerHidden?: boolean
  isSmallBusiness?: boolean
}

const sortBySequence = <T extends object>(a: T, b: T) => {
  return sortByObjectProperty('sequence' as keyof T, a, b)
}

class BasketTransformer {
  private basket: TransformedBasket = {}

  private getIsBannerHidden(basket: TransformedBasket): boolean {
    const { packages } = basket
    const isCombi = (bundle?: BasketV2.Bundle): boolean => {
      const offerGroup = get(bundle, 'priceDetails.merchandisingPromotions.offerGroup', '')
      return offerGroup?.toLowerCase() === 'combi offers' || offerGroup?.toLowerCase() === 'combined combi'
    }
    return packages ? packages.some(item => isCombi(item.bundle)) : false
  }

  private getIsBusiness(userState: UserState) {
    const customerSegment = getCookie('customerSegment')
    return customerSegment?.toLowerCase() === 'business' || userState?.accountSubCategory?.toLowerCase() === 'sole-trader'
  }

  private getIsSmallBusiness(userState: UserState) {
    return matchesConstCI(BASKET_CONSTS.ACCOUNT_TYPE_SMALLBUSINESS, userState.accountType)
  }

  public transformBasket(rawBasket: any, reviewMode?: boolean, userState: UserState = {}): TransformedBasket {
    const basket: TransformedBasket = sanitizeJson(rawBasket)

    basket.isReviewMode = reviewMode
    basket.isBusiness = this.getIsBusiness(userState)
    basket.isBannerHidden = this.getIsBannerHidden(basket)
    basket.isSmallBusiness = this.getIsSmallBusiness(userState)

    if (basket.packages) {
      basket.packages.sort(sortBySequence)
    }
    basket.isUpgradeOrder = isUpgradeJourney(basket.journey?.journeyType)

    this.basket = basket

    return this.basket
  }
}

export default new BasketTransformer()
