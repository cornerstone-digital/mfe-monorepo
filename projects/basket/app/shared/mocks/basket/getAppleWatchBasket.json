{
  "basketId": "4f50d9d5-e1db-4d19-b95c-c1f549cc95f5",
  "affiliateId": "11",
  "source": "eShop",
  "customerRequestedDate": "2017-10-01T09:45:00.000+02:00",
  "metadata": {
    "returnURL": null,
    "basketStatus": "INCOMPLETE",
    "salesOrderId": null,
    "reason": null
  },
  "journey": {
    "journeyType": "SecondLine",
    "offerCode": null
  },
  "packages": [
    {
      "packageId": "be925e3c-be5b-40b3-9cc0-dcc344a5c49f",
      "accountCategory": "Consumer",
      "sequence": 1,
      "packageType": "SecondLine",
      "removable": true,
      "dependentOn": [],
      "voucherCode": null,
      "bundle": {
        "packageLineId": "4afb17e2-9ddf-4119-8760-f1845d554e2a",
        "action": "ADD",
        "skuId": "111195",
        "name": "One Number Plan A",
        "description": "One Number Plan A",
        "displayName": "One Number Plan A",
        "displayDescription": null,
        "bundleClass": "HANDSET",
        "productClass": null,
        "serviceId": "110286",
        "paymentType": "Postpaid",
        "bundleType": "Red Extra",
        "commitmentPeriod": {
          "uom": "Months",
          "value": "24"
        },
        "displayOrder": null,
        "showInSummary": false,
        "allowances": [
          {
            "type": "UK MIN",
            "value": "Unlimited",
            "uom": "-",
            "displayUom": "minutes",
            "tilUom": "VOICE"
          },
          {
            "type": "UK TEXT",
            "value": "Unlimited",
            "uom": "-",
            "displayUom": "texts",
            "tilUom": "SMS"
          },
          {
            "type": "World Traveller",
            "value": "Vodafone Global Roaming",
            "uom": null,
            "displayUom": null,
            "tilUom": null
          }
        ],
        "merchandisingMedia": [],
        "attributes": null,
        "priceDetails": {
          "monthlyPrice": {
            "gross": "22",
            "net": "18.33",
            "vat": "3.67"
          },
          "monthlyDiscountPrice": {
            "gross": "22",
            "net": "18.33",
            "vat": "3.67"
          },
          "merchandisingPromotions": null,
          "listOfMerchandisingPromotion": [],
          "removedPromotionFootNote": null
        },
        "isVodafoneGlobalRoaming": true,
        "bundledServiceProducts": null,
        "planCategory": "CONSUMER",
        "portability": null,
        "isChanged": false
      },
      "hardwares": [
        {
          "packageLineId": "c61223ee-01f2-4d49-8bd4-9377571d7bf9",
          "action": "ADD",
          "skuId": "202040",
          "name": "Apple Watch Series 3 38mm silver",
          "description": null,
          "displayName": "Apple Watch 3 38mm",
          "displayDescription": null,
          "deviceBriefDescription": null,
          "deviceColor": null,
          "memorySize": null,
          "make": "apple",
          "model": "watch-3",
          "groupType": null,
          "productClass": "eHandset",
          "productSubClass": "WATCH",
          "displayOrder": null,
          "showInSummary": false,
          "merchandisingMedia": [],
          "attributes": null,
          "contractOptions": null,
          "priceDetails": {
            "oneOffPrice": {
              "gross": "0",
              "net": "0",
              "vat": "0"
            },
            "oneOffDiscountPrice": {
              "gross": "0",
              "net": "0",
              "vat": "0"
            },
            "merchandisingPromotions": null,
            "monthlyPrice": {
              "gross": null,
              "net": null,
              "vat": null
            },
            "monthlyDiscountPrice": {
              "gross": null,
              "net": null,
              "vat": null
            },
            "financingOptions": null,
            "deviceFinancingId": null,
            "devicePaymentPlan": null
          },
          "boxPrice": {
            "gross": "309.56",
            "net": "257.97",
            "vat": "51.59"
          },
          "freeGift": false,
          "isChanged": false
        }
      ],
      "services": [],
      "discounts": [],
      "priceDetails": {
        "oneOffPrice": {
          "gross": "0",
          "net": "0",
          "vat": "0"
        },
        "oneOffDiscountPrice": {
          "gross": "0",
          "net": "0",
          "vat": "0"
        },
        "monthlyPrice": {
          "gross": "22",
          "net": "18.33",
          "vat": "3.67"
        },
        "monthlyDiscountPrice": {
          "gross": "22",
          "net": "18.33",
          "vat": "3.67"
        },
        "financingOptions": null
      },
      "installationAddress": null,
      "appointmentWindow": null,
      "servicePoint": null,
      "serviceStartDate": null,
      "phoneNumber": null,
      "primaryDeviceIdentifier": "0789999999",
      "linePackageType": null,
      "lineTreatmentType": null,
      "planType": "EHANDSET",
      "combiPackageId": null,
      "deviceFinancingId": null,
      "isChanged": false
    }
  ],
  "vetOutcome": null,
  "priceDetails": {
    "oneOffPrice": {
      "gross": "0",
      "net": "0",
      "vat": "0"
    },
    "oneOffDiscountPrice": {
      "gross": "0",
      "net": "0",
      "vat": "0"
    },
    "monthlyPrice": {
      "gross": "22",
      "net": "18.33",
      "vat": "3.67"
    },
    "monthlyDiscountPrice": {
      "gross": "22",
      "net": "18.33",
      "vat": "3.67"
    }
  },
  "totalOneoffPriceSaving": "0.0",
  "totalMonthlyPriceSaving": "0",
  "deliveryInfo": null,
  "voucherCode": null,
  "voucherDescription": null,
  "validated": true,
  "hashvalue": "-901343189",
  "lastModified": null,
  "changeCode": null,
  "combiPackageIds": [],
  "voucherErrorMessage": null,
  "activeBundles": [
    {
      "packageLineId": "ee99f122-2356-423e-9a25-4122a13214cb",
      "bundleId": "110082",
      "identifier": {
        "key": "msisdn",
        "value": "447000000000"
      },
      "assetIntegrationId": "1-470I9UF",
      "activeDiscounts": [],
      "bundlePrice": {
        "bundleId": "110082",
        "monthlyPrice": {
          "gross": "51.79",
          "net": "43.16",
          "vat": "8.63"
        },
        "monthlyDiscountPrice": {
          "gross": "51.79",
          "net": "43.16",
          "vat": "8.63"
        },
        "merchandisingPromotions": null
      }
    }
  ],
  "splitDeliveryEligible": false,
  "priceProposal": [],
  "affiliateFlag": false,
  "isChanged": false
}