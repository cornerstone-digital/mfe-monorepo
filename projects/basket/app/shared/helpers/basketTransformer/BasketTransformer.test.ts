import BasketTransformer from './BasketTransformer'
import basket from '@basketMocks/basket.mock.json'

describe('BasketTransformer', () => {
  it('should transform the basket', () => {
    expect(BasketTransformer.transformBasket(basket, true, { accountSubCategory: 'soleTrader' })).toMatchSnapshot()
  })

  it('should have isSmallBusiness true in case of small business basket', () => {
    expect(BasketTransformer.transformBasket(basket, true, { accountType: 'Small-Business' })).toHaveProperty('isSmallBusiness', true)
  })
})
