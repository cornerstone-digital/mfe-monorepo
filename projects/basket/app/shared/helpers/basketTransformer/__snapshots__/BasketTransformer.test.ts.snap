// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasketTransformer should transform the basket 1`] = `
Object {
  "affiliateFlag": false,
  "basketId": "3a1ea657-3ee8-4375-84c8-929f5f14588c",
  "customerRequestedDate": "2018-10-03T10:48:13.000Z",
  "hashvalue": "965400216",
  "isBannerHidden": false,
  "isBusiness": false,
  "isChanged": false,
  "isReviewMode": true,
  "isSmallBusiness": false,
  "isUpgradeOrder": false,
  "metadata": Object {
    "basketStatus": "INCOMPLETE",
    "reason": "",
    "returnURL": "",
    "salesOrderId": "",
  },
  "packages": Array [
    Object {
      "accountCategory": "Consumer",
      "bundle": Object {
        "action": "ADD",
        "allowances": Array [
          Object {
            "displayUom": "GB",
            "tilUom": "DATA",
            "type": "UK DATA",
            "uom": "GB",
            "value": "8",
          },
          Object {
            "displayUom": "minutes",
            "tilUom": "VOICE",
            "type": "UK MIN",
            "uom": "-",
            "value": "Unlimited",
          },
          Object {
            "displayUom": "GB",
            "tilUom": "DATA",
            "type": "ORIGDATA",
            "uom": "GB",
            "value": "8",
          },
          Object {
            "displayUom": "minutes",
            "tilUom": "INTLVOICE",
            "type": "UK MIN",
            "uom": "-",
            "value": "500",
          },
          Object {
            "displayUom": "texts",
            "tilUom": "SMS",
            "type": "UK TEXT",
            "uom": "-",
            "value": "Unlimited",
          },
        ],
        "bundleClass": "HANDSET",
        "bundleType": "Red Entertainment",
        "bundledServiceProducts": Array [],
        "commitmentPeriod": Object {
          "uom": "Months",
          "value": "18",
        },
        "description": "Red Entertainment Plan comes with unlimited mins, unlimited texts, 8GB data, 500m to EU & VGRPlus",
        "displayDescription": "",
        "displayName": "8GB Red Entertainment plan",
        "isChanged": false,
        "isVodafoneGlobalRoaming": true,
        "merchandisingMedia": Array [
          Object {
            "footNotes": Array [
              "",
            ],
            "id": "entertainment.merchandisingPromotions.merchandisingPromotion.PromotionMedia",
            "type": "URL",
            "value": "/images/desktop/img_mef_ent_packs.png",
          },
          Object {
            "footNotes": Array [
              "",
            ],
            "id": "securenet.merchandisingPromotions.merchandisingPromotion.label",
            "type": "TEXT",
            "value": "3-month free trial of Secure Net",
          },
          Object {
            "footNotes": Array [
              "",
            ],
            "id": "entertainment.merchandisingPromotions.merchandisingPromotion.label",
            "type": "TEXT",
            "value": "Your choice of entertainment for 24 months",
          },
        ],
        "name": "24mth Band E Unlmin 8GB Red Entertainment Plan with 500m to EU and VGRPlus",
        "packageLineId": "ecc44220-186a-4aab-a6bc-f236d273b131",
        "paymentType": "Postpaid",
        "planCategory": "",
        "priceDetails": Object {
          "merchandisingPromotions": Object {},
          "monthlyDiscountPrice": Object {
            "gross": "51",
            "net": "42.50",
            "vat": "8.50",
          },
          "monthlyPrice": Object {
            "gross": "51",
            "net": "42.50",
            "vat": "8.50",
          },
        },
        "productClass": "",
        "serviceId": "110286",
        "showInSummary": false,
        "skuId": "111079",
      },
      "dependentOn": Array [
        "",
      ],
      "hardwares": Array [
        Object {
          "action": "ADD",
          "boxPrice": Object {
            "gross": "0.0",
            "net": "0.0",
            "vat": "0.0",
          },
          "description": "",
          "deviceBriefDescription": Array [],
          "deviceColor": "",
          "displayDescription": "",
          "displayName": "3-in-1 SIM",
          "freeGift": false,
          "groupType": "",
          "isChanged": false,
          "make": "",
          "memorySize": "",
          "merchandisingMedia": Array [],
          "model": "",
          "name": "Blank White Triple Format SIM",
          "packageLineId": "b8158d37-5e98-4550-833b-094d1b9a3d18",
          "priceDetails": Object {
            "oneOffDiscountPrice": Object {
              "gross": "0",
              "net": "0",
              "vat": "0",
            },
            "oneOffPrice": Object {
              "gross": "0",
              "net": "0",
              "vat": "0",
            },
          },
          "productClass": "SIM Card",
          "showInSummary": false,
          "skuId": "086532",
        },
        Object {
          "action": "ADD",
          "boxPrice": Object {
            "gross": "432.0",
            "net": "360.0",
            "vat": "72.0",
          },
          "description": "Boundary-breaking design, check. Infinity Display, check. Personal assistant in your pocket? Check. Samsung S8 has it all.",
          "deviceBriefDescription": Array [],
          "deviceColor": "Orchid Grey",
          "displayDescription": "",
          "displayName": "Samsung Galaxy S8 64GB Orchid Grey",
          "freeGift": false,
          "groupType": "DEVICE_PAYM",
          "isChanged": false,
          "make": "samsung",
          "memorySize": "64 GB",
          "merchandisingMedia": Array [
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.thumbs.back",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-back.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.grid",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-grid-product-front.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.thumbs.left",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-left.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.right",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-right.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.hero",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-front.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.thumbs.front",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-front.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.thumbs.side",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-side.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.left",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-left.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.back",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-back.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.front",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-front.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.side",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-side.png",
            },
            Object {
              "id": "imageURLs.thumbs.right",
              "planCategory": Array [
                "",
              ],
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-right.png",
            },
          ],
          "model": "galaxy-s8",
          "name": "Samsung Galaxy S8 orchid grey",
          "packageLineId": "7bfe074b-bab4-4e90-9145-782afbb3fc41",
          "priceDetails": Object {
            "oneOffDiscountPrice": Object {
              "gross": "9",
              "net": "7.50",
              "vat": "1.50",
            },
            "oneOffPrice": Object {
              "gross": "9",
              "net": "7.50",
              "vat": "1.50",
            },
          },
          "productClass": "Handset",
          "showInSummary": false,
          "skuId": "090357",
        },
      ],
      "isChanged": false,
      "packageId": "fee83b9d-d084-4ae3-a198-9ff630665159",
      "packageType": "Acquisition",
      "planType": "SIMO",
      "priceDetails": Object {
        "monthlyDiscountPrice": Object {
          "gross": "51",
          "net": "42.50",
          "vat": "8.50",
        },
        "monthlyPrice": Object {
          "gross": "51",
          "net": "42.50",
          "vat": "8.50",
        },
        "oneOffDiscountPrice": Object {
          "gross": "9",
          "net": "7.50",
          "vat": "1.50",
        },
        "oneOffPrice": Object {
          "gross": "9",
          "net": "7.50",
          "vat": "1.50",
        },
      },
      "removable": true,
      "sequence": 1,
      "services": Array [
        Object {
          "action": "ADD",
          "dataRollover": "",
          "description": "",
          "displayDescription": "",
          "displayName": "Loss, theft, damage and breakdown cover",
          "freeGift": false,
          "isChanged": false,
          "merchandisingMedia": Array [],
          "name": "Vodafone Insurance: Tier 1 Loss Theft and Damage",
          "packageLineId": "17e76d94-44b7-4e00-935a-e2e0b283cdfa",
          "priceDetails": Object {
            "monthlyDiscountPrice": Object {
              "gross": "13.50",
              "net": "13.50",
              "vat": "0",
            },
            "monthlyPrice": Object {
              "gross": "13.50",
              "net": "13.50",
              "vat": "0",
            },
          },
          "productClass": "Insurance",
          "rewardPoints": "",
          "rootProductId": "110461",
          "skuId": "110461",
          "specificationGroup": Array [],
        },
      ],
      "tradeInCredit": Object {
        "credit": Object {
          "creditProductId": 94568,
          "guaranteedPrice": 380,
          "type": "BACS",
        },
        "deviceId": "16096626-c46d-4caa-8dxx-2ce5b2516455",
        "deviceName": "Apple iPhone11pro 256Gb black",
        "diagnostic": true,
        "expiryDateTime": "2022-05-20T09:09:48",
        "grade": Object {
          "condition": "working",
        },
        "quoteId": "15670678-b46d-4cef-8d3a-2ce5b2516c63",
      },
      "tradeInOfferCode": "Tier2",
      "voucherCode": "",
    },
    Object {
      "accountCategory": "Consumer",
      "bundle": Object {
        "action": "ADD",
        "allowances": Array [
          Object {
            "displayUom": "GB",
            "tilUom": "DATA",
            "type": "UK DATA",
            "uom": "GB",
            "value": "8",
          },
          Object {
            "displayUom": "minutes",
            "tilUom": "VOICE",
            "type": "UK MIN",
            "uom": "-",
            "value": "Unlimited",
          },
          Object {
            "displayUom": "GB",
            "tilUom": "DATA",
            "type": "ORIGDATA",
            "uom": "GB",
            "value": "8",
          },
          Object {
            "displayUom": "minutes",
            "tilUom": "INTLVOICE",
            "type": "UK MIN",
            "uom": "-",
            "value": "500",
          },
          Object {
            "displayUom": "texts",
            "tilUom": "SMS",
            "type": "UK TEXT",
            "uom": "-",
            "value": "Unlimited",
          },
        ],
        "bundleClass": "HANDSET",
        "bundleType": "Red Entertainment",
        "bundledServiceProducts": Array [],
        "commitmentPeriod": Object {
          "uom": "Months",
          "value": "18",
        },
        "description": "Red Entertainment Plan comes with unlimited mins, unlimited texts, 8GB data, 500m to EU & VGRPlus",
        "displayDescription": "",
        "displayName": "8GB Red Entertainment plan",
        "isChanged": false,
        "isVodafoneGlobalRoaming": true,
        "merchandisingMedia": Array [
          Object {
            "footNotes": Array [
              "",
            ],
            "id": "entertainment.merchandisingPromotions.merchandisingPromotion.PromotionMedia",
            "type": "URL",
            "value": "/images/desktop/img_mef_ent_packs.png",
          },
          Object {
            "footNotes": Array [
              "",
            ],
            "id": "securenet.merchandisingPromotions.merchandisingPromotion.label",
            "type": "TEXT",
            "value": "3-month free trial of Secure Net",
          },
          Object {
            "footNotes": Array [
              "",
            ],
            "id": "entertainment.merchandisingPromotions.merchandisingPromotion.label",
            "type": "TEXT",
            "value": "Your choice of entertainment for 24 months",
          },
        ],
        "name": "24mth Band E Unlmin 8GB Red Entertainment Plan with 500m to EU and VGRPlus",
        "packageLineId": "ecc44220-186a-4aab-a6bc-f236d273b131",
        "paymentType": "Postpaid",
        "planCategory": "",
        "priceDetails": Object {
          "monthlyDiscountPrice": Object {
            "gross": "51",
            "net": "42.50",
            "vat": "8.50",
          },
          "monthlyPrice": Object {
            "gross": "51",
            "net": "42.50",
            "vat": "8.50",
          },
        },
        "productClass": "",
        "serviceId": "110286",
        "showInSummary": false,
        "skuId": "111079",
      },
      "dependentOn": Array [
        "",
      ],
      "hardwares": Array [
        Object {
          "action": "ADD",
          "boxPrice": Object {
            "gross": "0.0",
            "net": "0.0",
            "vat": "0.0",
          },
          "description": "",
          "deviceBriefDescription": Array [],
          "deviceColor": "",
          "displayDescription": "",
          "displayName": "3-in-1 SIM",
          "freeGift": false,
          "groupType": "",
          "isChanged": false,
          "make": "",
          "memorySize": "",
          "merchandisingMedia": Array [],
          "model": "",
          "name": "Blank White Triple Format SIM",
          "packageLineId": "b8158d37-5e98-4550-833b-094d1b9a3d18",
          "priceDetails": Object {
            "oneOffDiscountPrice": Object {
              "gross": "0",
              "net": "0",
              "vat": "0",
            },
            "oneOffPrice": Object {
              "gross": "0",
              "net": "0",
              "vat": "0",
            },
          },
          "productClass": "SIM Card",
          "showInSummary": false,
          "skuId": "086532",
        },
        Object {
          "action": "ADD",
          "boxPrice": Object {
            "gross": "432.0",
            "net": "360.0",
            "vat": "72.0",
          },
          "description": "Boundary-breaking design, check. Infinity Display, check. Personal assistant in your pocket? Check. Samsung S8 has it all.",
          "deviceBriefDescription": Array [],
          "deviceColor": "Orchid Grey",
          "displayDescription": "",
          "displayName": "Samsung Galaxy S8 64GB Orchid Grey",
          "freeGift": false,
          "groupType": "DEVICE_PAYM",
          "isChanged": false,
          "make": "samsung",
          "memorySize": "64 GB",
          "merchandisingMedia": Array [
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.thumbs.back",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-back.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.grid",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-grid-product-front.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.thumbs.left",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-left.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.right",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-right.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.hero",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-front.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.thumbs.front",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-front.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.thumbs.side",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-side.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.left",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-left.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.back",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-back.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.front",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-front.png",
            },
            Object {
              "footNotes": Array [
                "",
              ],
              "id": "imageURLs.full.side",
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-full-product-side.png",
            },
            Object {
              "id": "imageURLs.thumbs.right",
              "planCategory": Array [
                "",
              ],
              "type": "URL",
              "value": "/images/desktop/Samsung_Galaxy_s8_orchid_grey-thumbs-right.png",
            },
          ],
          "model": "galaxy-s8",
          "name": "Samsung Galaxy S8 orchid grey",
          "packageLineId": "7bfe074b-bab4-4e90-9145-782afbb3fc41",
          "priceDetails": Object {
            "oneOffDiscountPrice": Object {
              "gross": "9",
              "net": "7.50",
              "vat": "1.50",
            },
            "oneOffPrice": Object {
              "gross": "9",
              "net": "7.50",
              "vat": "1.50",
            },
          },
          "productClass": "Handset",
          "showInSummary": false,
          "skuId": "090357",
        },
      ],
      "isChanged": false,
      "packageId": "20434cf7-be99-4e9d-878d-b9bac8f83072",
      "packageType": "Acquisition",
      "planType": "BroadBand:FTTC",
      "priceDetails": Object {
        "monthlyDiscountPrice": Object {
          "gross": "51",
          "net": "42.50",
          "vat": "8.50",
        },
        "monthlyPrice": Object {
          "gross": "51",
          "net": "42.50",
          "vat": "8.50",
        },
        "oneOffDiscountPrice": Object {
          "gross": "9",
          "net": "7.50",
          "vat": "1.50",
        },
        "oneOffPrice": Object {
          "gross": "9",
          "net": "7.50",
          "vat": "1.50",
        },
      },
      "removable": true,
      "sequence": 1,
      "services": Array [
        Object {
          "action": "ADD",
          "dataRollover": "",
          "description": "",
          "displayDescription": "",
          "displayName": "Loss, theft, damage and breakdown cover",
          "freeGift": false,
          "isChanged": false,
          "merchandisingMedia": Array [],
          "name": "Vodafone Insurance: Tier 1 Loss Theft and Damage",
          "packageLineId": "17e76d94-44b7-4e00-935a-e2e0b283cdfa",
          "priceDetails": Object {
            "monthlyDiscountPrice": Object {
              "gross": "13.50",
              "net": "13.50",
              "vat": "0",
            },
            "monthlyPrice": Object {
              "gross": "13.50",
              "net": "13.50",
              "vat": "0",
            },
          },
          "productClass": "Insurance",
          "rewardPoints": "",
          "rootProductId": "110461",
          "skuId": "110461",
          "specificationGroup": Array [],
        },
      ],
      "voucherCode": "",
    },
  ],
  "priceDetails": Object {
    "monthlyDiscountPrice": Object {
      "gross": "51",
      "net": "42.50",
      "vat": "8.50",
    },
    "monthlyPrice": Object {
      "gross": "51",
      "net": "42.50",
      "vat": "8.50",
    },
    "oneOffDiscountPrice": Object {
      "gross": "9",
      "net": "7.50",
      "vat": "1.50",
    },
    "oneOffPrice": Object {
      "gross": "9",
      "net": "7.50",
      "vat": "1.50",
    },
  },
  "source": "WCS",
  "totalMonthlyPriceSaving": "0",
  "totalOneoffPriceSaving": "0.0",
  "validated": false,
}
`;
